src/python/pants/jvm/classpath.py

```diff
@@ -31,33 +30,39 @@ class Classpath(Collection[ClasspathEntry]):
     This classpath is guaranteed to contain only JAR files.
     """
 
+    def __init__(self, entries: tuple[ClasspathEntry, ...], resolve: CoursierResolveKey) -> None:
+        self.entries: tuple[ClasspathEntry, ...] = entries
+        self.resolve: CoursierResolveKey = resolve
```

Comment:
This no longer has useful `eq`/`hash`, which will prevent useful memoization: should annotate as `@dataclass`, and can then replace your `__init__` with:
```suggestion
    entries: tuple[ClasspathEntry, ...]
    resolve: CoursierResolveKey
```